managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1033,26 +1041,81 @@ public void asyncReadEntriesWithSkipOrWait(int maxEntries, long maxSizeBytes, Re
             OpReadEntry op = OpReadEntry.create(this, readPosition, numberOfEntriesToRead, callback,
                     ctx, maxPosition, skipCondition);
 
-            if (!WAITING_READ_OP_UPDATER.compareAndSet(this, null, op)) {
-                op.recycle();
-                callback.readEntriesFailed(new ManagedLedgerException.ConcurrentWaitCallbackException(), ctx);
-                return;
-            }
+            synchronized (pendingReadOpMutex) {
+                if (!WAITING_READ_OP_UPDATER.compareAndSet(this, null, op)) {
+                    op.recycle();
+                    callback.readEntriesFailed(new ManagedLedgerException.ConcurrentWaitCallbackException(), ctx);
+                    return;
+                }
 
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] [{}] Deferring retry of read at position {}", ledger.getName(), name, op.readPosition);
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] [{}] Deferring retry of read at position {}", ledger.getName(), name,
+                            op.readPosition);
+                }
+
+                // Check again for new entries after the configured time, then if still no entries are available register
+                // to be notified
+                if (getConfig().getNewEntriesCheckDelayInMillis() > 0) {
+                    delayCheckForNewEntriesTask = new DelayCheckForNewEntriesTask(op, callback, ctx);
+                } else {
+                    // If there's no delay, check directly from the same thread
+                    checkForNewEntries(op, callback, ctx);
+                }
             }
+        }
+    }
 
-            // Check again for new entries after the configured time, then if still no entries are available register
-            // to be notified
-            if (getConfig().getNewEntriesCheckDelayInMillis() > 0) {
-                ledger.getScheduledExecutor()
-                        .schedule(() -> checkForNewEntries(op, callback, ctx),
-                                getConfig().getNewEntriesCheckDelayInMillis(), TimeUnit.MILLISECONDS);
-            } else {
-                // If there's no delay, check directly from the same thread
+    private class DelayCheckForNewEntriesTask implements Runnable {
+
+        private final OpReadEntry op;
+        private final ReadEntriesCallback callback;
+        private final Object ctx;
+        private ScheduledFuture<?> scheduledFuture;
+        /**
+         * 0: init
+         * 1: running
+         * 2: cancelled
+         * 3: done
+         */
```

Comment:
Please don't use magic values in code.
Replace with constants or enums. (In Java, all Enums have integer values with `.ordinal()`)

Reply:
Addressed